### PR TITLE
Implementation of asynchronous integration API.

### PR DESCRIPTION
**Pull request by halfflat**
* Stage events for next integration interval on lowered cell.
* Use explicit binning for event coalescence.
* Extend `event_queue` to allow checking top of queue against arbitrary predicates.
* Add `--bin-dt` and `--bin-regular` options to miniapp (disable binning with `--bin-dt 0`).
* Tidy up miniapp option settings class.

Integration in lowered cell over multiple steps is deferred until samplers can be set up with back-end polling.

Asynchronous integration itself is not yet implemented.

### CODE REVIEW

**Comment by w-klijn** [comment]
Would it be smart to rename the merge? something like prepare for asynch..

Just because the last line of the description invalidates the title completely.

File: src/cell_group.hpp [hunk 1]
```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
```
Comment (w-klijn) [inline src/cell_group.hpp]: This function is getting very long: Would it be smart to split it in multiple steps to improve readability?

File: src/cell_group.hpp [hunk 2]
```diff
@@ -231,7 +241,7 @@ class cell_group {
         auto handle = get_probe_handle(probe_id);
 
         auto sampler_index = uint32_t(samplers_.size());
-        samplers_.push_back({handle, s});
+        samplers_.push_back({handle, probe_id.gid, s});
```
Comment (w-klijn) [inline src/cell_group.hpp]: The interface of this constructor would be nicer if only the probe_id would be given (instead of the gid).

File: src/event_queue.hpp [hunk 3]
```diff
@@ -85,6 +103,22 @@ public :
...
+    // Pop and return top event `ev` of queue if `t_until` > `event_time(ev)`.
+    util::optional<value_type> pop_if_before(const time_type& t_until) {
+        using ::nest::mc::event_time;
+        return pop_if(
+            [&t_until](const value_type& ev) { return t_until > event_time(ev); }
+        );
+    }
+
+    // Pop and return top event `ev` of queue unless `event_time(ev)` > `t_until`
+    util::optional<value_type> pop_if_not_after(const time_type& t_until) {
```
Comment (w-klijn) [inline src/event_queue.hpp]: negation in function names can be confusing. Is there a way to rename it?

Comment (halfflat) [inline src/event_queue.hpp]: Here I think it is unavoidable; because we're limiting ourselves to the use of a strict inequality `>` on time types, we're making a distinction between 'before' (i.e. time is less than top of queue) and 'not after' (i.e. time is not greater than top of queue).

An alternative, `pop_if_before_or_same`, would imply we're testing for equality or greater-than-equal-to, when in truth we're still just using greater-than.

File: src/cell_group.hpp (hunk 1 above)
Comment (halfflat) [inline src/cell_group.hpp]: Ignoring the extra comments, the function is pretty much the same length as before.

That said, it will be much simpler when sampling is also pushed to the back-end, so I would defer a refactoring of this method until then.

File: src/cell_group.hpp (hunk 2 above)
Comment (halfflat) [inline src/cell_group.hpp]: We only need the gid though…

Comment (w-klijn) [inline src/cell_group.hpp]: With this implementation the cell_group knows 'that the probe has a gid'.  tight coupling

File: src/cell_group.hpp [hunk 4]
```diff
@@ -143,51 +143,61 @@ class cell_group {
...
+
+        // Bin pending events and enqueue on lowered state.
+        time_type ev_min_time = lowered_.max_time(); // (but we're synchronized here)
+        while (auto ev = events_.pop_if_before(tfinal)) {
+            auto handle = get_target_handle(ev->target);
+            auto binned_ev_time = binner_.bin(ev->target.gid, ev->time, ev_min_time);
+            lowered_.add_event(binned_ev_time, handle, ev->weight);
+        }
+
+        lowered_.start_integration(tfinal, dt);
```
Comment (bcumming) [inline src/cell_group.hpp]: The name is a bit misleading, maybe `prepare_integration` or `setup_integration` are more appropriate.

If I understand the comments and your plans correctly, the event delivery and sampler polling will both be moved to the lowered cells, and thus be able to perform time stepping to `t_final` with one call?

In that case, would we need to have a separate `start_integration` and `step_integration` methods?

File: src/cell_group.hpp (hunk 1 above)
Comment (bcumming) [inline src/cell_group.hpp]: @w-klijn is right, in that this function is a bit hard to follow. Sometimes a longer function can make complicated code easier to understand, but in this case things like the sampling logic could be refactored.

But this is a work in progress, it should be cleaner once much of the event and sampling logic has been moved to the back end.

File: src/cell_group.hpp [hunk 5]
```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
+
+        // Bin pending events and enqueue on lowered state.
```
Comment (bcumming) [inline src/cell_group.hpp]: Is this covered by the profiler? Try to make sure that the profiler is catching everything once the refactoring is complete.

File: src/fvm_multicell.hpp [hunk 6]
```diff
@@ -85,8 +85,56 @@ class fvm_multicell {
         return (this->*h.first)[h.second];
     }
 
-    /// integrate all cell state forward in time
-    void advance(double dt);
+    /// Initialize state prior to a sequence of integration steps.
+    void start_integration(value_type tfinal, value_type dt_max) {
```
Comment (bcumming) [inline src/fvm_multicell.hpp]: The name `start_integration` gives the impression that integration will be performed when called.

File: src/fvm_multicell.hpp [hunk 7]
```diff
@@ -85,8 +85,56 @@ class fvm_multicell {
...
+
+    /// Advance one integration step.
+    void step_integration();
+
+    /// Query integration completion state.
+    bool integration_complete() const {
+        return !integration_running_;
+    }
+
+    /// Query per-cell time state. (Placeholders)
```
Comment (bcumming) [inline src/fvm_multicell.hpp]: Placeholders for what? Now would be a good time to have more detailed comments that explain the meaning of each returned value, to make it easier for me to see where this is going.

File: src/fvm_multicell.hpp [hunk 8]
```diff
@@ -85,8 +85,56 @@ class fvm_multicell {
...
+        EXPECTS(events_.empty());
+        for (auto& ev: staged_events_) {
+            EXPECTS(ev.time<tfinal);
+            events_.push(std::move(ev));
+        }
+
+        staged_events_.clear();
+    }
+
+    /// Advance one integration step.
```
Comment (bcumming) [inline src/fvm_multicell.hpp]: `// this is a loop`
sorry.

File: src/fvm_multicell.hpp [hunk 9]
```diff
@@ -85,8 +85,56 @@ class fvm_multicell {
...
+
+    value_type max_time() const {
+        return t_;
+    }
+
+    bool state_synchronized() const {
+        return true;
+    }
+
+    /// Add an event for processing in next integration stage.
```
Comment (bcumming) [inline src/fvm_multicell.hpp]: Do the events have to be processed in the next integration state? Is it true that an event that will occur in a later integration stage could be added, and will be persistent in the queue over multiple stages, until it is time to deliver it?

File: src/fvm_multicell.hpp [hunk 10]
```diff
@@ -194,15 +240,38 @@ class fvm_multicell {
...
-    value_type t_ = value_type{0};
+    value_type t_ = 0;
 
     /// resting potential (initial voltage condition)
     value_type resting_potential_ = -65;
 
+    /// final time in integration round [ms]
+    value_type tfinal_ = 0;
+
+    /// max time step for integration [ms]
```
Comment (bcumming) [inline src/fvm_multicell.hpp]: A quibble about naming: this could also be referred to as the `target_dt`, i.e. the time step size that we would like to take, events and `t_final` permitting. The comment might want to mention this.

File: src/fvm_multicell.hpp [hunk 11]
```diff
@@ -194,15 +240,38 @@ class fvm_multicell {
...
+    /// once integration to `tfinal_` is complete.
+    bool integration_running_ = false;
+
+    struct deliverable_event {
+        double time;
+        target_handle handle;
+        double weight;
+    };
+
+    /// events staged for upcoming integration stage
```
Comment (bcumming) [inline src/fvm_multicell.hpp]: see comment about: choose to insert either `an` or `the` in front of `upcoming` in this comment.

File: src/cell_group.hpp (hunk 2 above)
Comment (bcumming) [inline src/cell_group.hpp]: Why the additional parameter?

Comment (halfflat) [inline src/cell_group.hpp]: The cell group needs to know the gid associated with the probe, so it can test the sampler's requested time against the time for that gid. It is a consequence of the asynchronous behaviour.

File: miniapp/io.cpp [hunk 12]
```diff
@@ -179,6 +146,11 @@ cl_options read_options(int argc, char** argv, bool allow_write) {
         TCLAP::ValueArg<double> dt_arg(
             "d", "dt", "set simulation time step to <time> ms",
             false, defopts.dt, "time", cmd);
+        TCLAP::ValueArg<double> bin_dt_arg(
+            "", "bin-dt", "set event binning interval to <time> ms",
+            false, defopts.bin_dt, "time", cmd);
+        TCLAP::SwitchArg bin_regular_arg(
+            "","bin-regular","use 'regular' binning policy instead of 'following'", cmd, false);
```
Comment (halfflat) [inline miniapp/io.cpp]: Totally want to rip out our option handling infrastructure and replace. _I have a plan_.

In the meantime, the hypens-in-long-options style is pretty common (it is certainly GNU style). Google test's underscore use is not standard practice. I think we should stick with it. We could modify the JSON keys to match, when we get around to this clean up.

File: src/cell_group.hpp (hunk 4 above)
Comment (halfflat) [inline src/cell_group.hpp]: Happy to change to `init_integration` or similar. It does, in some sense, get things going: 'prepare' would be a bit misleading.

Either way though, indeed once sampling is also performed at the lowered level there will be no need to have separate methods here.

File: src/fvm_multicell.hpp (hunk 8 above)
Comment (halfflat) [inline src/fvm_multicell.hpp]: I didn't think it was really that transparent — didn't want to use `advance` in the method name to avoid confusion with the previous method and semantics, and I thought the name in isolation was a little ambiguous.

File: src/fvm_multicell.hpp (hunk 9 above)
Comment (halfflat) [inline src/fvm_multicell.hpp]: The assumption here is no: all the events for one 'big' interval and no others are expected to be passed on to the lowered cell. It may be possible to relax this assumption, but I wouldn't do so unless there is a clear gain.

The `cell_group` is responsible for queuing, and holds the priority queue data structure. The lowered implementation now has responsibility for delivering, and the API is meant to enforce a sequence of:
1. `cell_group` provides all the events, associated with their gid.
2. `cell_group` instructs lowered implementation to bulk-load/prepare the provided events for delivery within the integration interval.
3. The lowered implementation uses a pop/top-only queue for handling the discrete deliveries internally over the course of processing an integration interval.

File: src/fvm_multicell.hpp (hunk 10 above)
Comment (halfflat) [inline src/fvm_multicell.hpp]: `dt_max_` is closer to the meaning associated with the interface (i.e. the `cell_group` requests a maximum _dt_). A `target_dt` is closer to the meaning as provided by this lowered cell implementation, as you describe.

A comment on how `dt_max_` is satisfied is more appropriate for the integration code, and I can add one there.

**Comment by halfflat** [comment]
I will update the corresponding ticket to outline the plan of PRs, with the hope of making the goal of these patches more transparent.

**Comment by halfflat** [comment]
Made a small update with a change of `start_integration` to `setup_integration`, and a comment regarding `dt_max_`.

I've updated issue #184 with a more detailed implementation plan.

File: miniapp/io.cpp (hunk 12 above)
Comment (bcumming) [inline miniapp/io.cpp]: I prefer the hyphen approach too, and second the motion to use hyphens in the json files.

**Comment by bcumming** [comment]
Looks good to me. On the condition that the "placeholder" comments turn into documentation in the final version.
